cpp/src/fil/fil.cu

```diff
@@ -781,19 +787,36 @@ template void init_sparse<sparse_node8>(const raft::handle_t& h, forest_t* pf,
                                         const sparse_node8* nodes,
                                         const forest_params_t* params);
 
+template <typename T, typename L>
 void from_treelite(const raft::handle_t& handle, forest_t* pforest,
-                   ModelHandle model, const treelite_params_t* tl_params) {
+                   const tl::ModelImpl<T, L>& model,
+                   const treelite_params_t* tl_params) {
+  // Invariants on threshold and leaf types
+  static_assert(std::is_same<T, float>::value || std::is_same<T, double>::value,
+                "Model must contain float32 or float64 thresholds for splits");
+  ASSERT((std::is_same<L, float>::value || std::is_same<L, double>::value),
+         "Models with integer leaf output are not yet supported");
+  // Display appropriate warnings when float64 values are being casted into
+  // float32, as FIL only supports inferencing with float32 for the time being
+  if (std::is_same<T, double>::value || std::is_same<L, double>::value) {
+    CUML_LOG_WARN(
+      "The tree ensemble model contains float64 values. Since FIL "
+      "currently only supports inferencing with float32 values, "
+      "all thresholds and leaf values will be converted from "
+      "float64 to float32. This may lead to inaccurate "
+      "prediction.");
+  }
```

Comment:
Note to reviewer: We finally have a robust method for warning about float64 use.